lock-impl/src/main/java/com/palantir/lock/impl/LockRefreshingTimelockService.java

```diff
@@ -0,0 +1,104 @@
+ * limitations under the License.
+ */
+
+package com.palantir.lock.impl;
+
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.palantir.lock.v2.LockImmutableTimestampRequest;
+import com.palantir.lock.v2.LockImmutableTimestampResponse;
+import com.palantir.lock.v2.LockRequestV2;
+import com.palantir.lock.v2.LockTokenV2;
+import com.palantir.lock.v2.TimelockService;
+import com.palantir.lock.v2.WaitForLocksRequest;
+import com.palantir.timestamp.TimestampRange;
+
+// TODO(nziebart): probably should make it more obvious that this class should always be used;
+// maybe call this a TimelockClient and require that everywhere? Could also be used for async unlocking..
+public class LockRefreshingTimelockService implements TimelockService {
+
+    private static final long REFRESH_INTERVAL_MILLIS = 5_000;
+
+    private final TimelockService delegate;
+    private final LockRefresher lockRefresher;
+
+    public static LockRefreshingTimelockService createDefault(TimelockService timelockService) {
+        ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor(new ThreadFactoryBuilder()
+                .setNameFormat(LockRefreshingTimelockService.class.getSimpleName() + "-%d")
+                .setDaemon(true)
+                .build());
+        LockRefresher lockRefresher = new LockRefresher(executor, timelockService, REFRESH_INTERVAL_MILLIS);
+        return new LockRefreshingTimelockService(timelockService, lockRefresher);
+    }
+
+    public LockRefreshingTimelockService(TimelockService delegate, LockRefresher lockRefresher) {
+        this.delegate = delegate;
+        this.lockRefresher = lockRefresher;
+    }
+
+    @Override
+    public long getFreshTimestamp() {
+        return delegate.getFreshTimestamp();
+    }
+
+    @Override
+    public TimestampRange getFreshTimestamps(int numTimestampsRequested) {
+        return delegate.getFreshTimestamps(numTimestampsRequested);
+    }
+
+    @Override
+    public LockImmutableTimestampResponse lockImmutableTimestamp(LockImmutableTimestampRequest request) {
+        LockImmutableTimestampResponse response = delegate.lockImmutableTimestamp(request);
+        lockRefresher.registerLock(response.getLock());
+        return response;
+    }
+
+    @Override
+    public long getImmutableTimestamp() {
+        return delegate.getImmutableTimestamp();
+    }
+
+    @Override
+    public LockTokenV2 lock(LockRequestV2 request) {
+        LockTokenV2 response = delegate.lock(request);
+        lockRefresher.registerLock(response);
+        return response;
+    }
+
+    @Override
+    public void waitForLocks(WaitForLocksRequest request) {
+        delegate.waitForLocks(request);
+    }
+
+    @Override
+    public Set<LockTokenV2> refreshLockLeases(Set<LockTokenV2> tokens) {
+        return delegate.refreshLockLeases(tokens);
```

Comment:
Yep, that was what I had in mind. Provided that all tokens are registered before they can be refreshed, this is safe to do. We could also do it for unlock. The only issue is that someone might extend `LockRefreshingTimelockService` and override `lock` or `lockImmutableTimestamp` so that it doesn't register the lock, but I think it's fine if we just write some javadocs pointing out the dangers of doing that (i.e. you **must** override the refresh/unlock accordingly).